src/solver/solvers/solver_ppc.h

```diff
@@ -132,151 +182,335 @@ namespace solver
+                                                     Messenger& fromCoarser,
+                                                     const double currentTime, const double newTime)
+    {
+        auto& hybridState      = model.state;
+        auto& resourcesManager = model.resourcesManager;
+        auto dt                = newTime - currentTime;
+        auto& electromag       = hybridState.electromag;
+        {
+            auto& Epred = electromagPred_.E;
+            auto& Bpred = electromagPred_.B;
+            auto& Bavg  = electromagAvg_.B;
+            auto& Eavg  = electromagAvg_.E;
+            auto& B     = hybridState.electromag.B;
+            auto& E     = hybridState.electromag.E;
+
+            for (auto& patch : level)
+            {
+                auto _ = resourcesManager->setOnPatch(*patch, electromagAvg_, electromagPred_,
+                                                      hybridState.electromag);
+                PHARE::core::average(B, Bpred, Bavg);
+                PHARE::core::average(E, Epred, Eavg);
+            }
+        }
+    }
+
+
+
+    template<typename HybridModel, typename AMR_Types>
+    void SolverPPC<HybridModel, AMR_Types>::moveIons_(level_t& level, Ions& ions,
+                                                      Electromag& electromag, ResourcesManager& rm,
+                                                      Messenger& fromCoarser,
+                                                      const double currentTime,
+                                                      const double newTime, core::UpdaterMode mode)
+    {
+        std::size_t nbrParticles = 0;
+        for (auto& patch : level)
+        {
+            auto _ = rm.setOnPatch(*patch, ions);
+
+            for (auto& pop : ions)
+            {
+                nbrParticles += pop.domainParticles().size();
+            }
+        }
+        std::cout << "t = " << currentTime << ";  nbr Particles = " << nbrParticles << "\n";
+
+
+
+
+        auto dt = newTime - currentTime;
 
+        for (auto& patch : level)
+        {
+            auto _ = rm.setOnPatch(*patch, electromag, ions);
+
+            auto layout = PHARE::amr::layoutFromPatch<GridLayout>(*patch);
+            ionUpdater_.updatePopulations(ions, electromag, layout, dt, mode);
+
+            // this needs to be done before calling the messenger
+            rm.setTime(ions, *patch, newTime);
+        }
+
+
+        fromCoarser.fillIonGhostParticles(ions, level, newTime);
+        fromCoarser.fillIonMomentGhosts(ions, level, currentTime, newTime);
+
+        for (auto& patch : level)
+        {
+            auto _      = rm.setOnPatch(*patch, electromag, ions);
+            auto layout = PHARE::amr::layoutFromPatch<GridLayout>(*patch);
+            ionUpdater_.updateIons(ions, layout);
+
+            // no need to update time, since it has been done before
+        }
+    }
+} // namespace solver
```

Comment:
cosmetic:

less indentation with

```
namespace PHARE::solver {
```

if you feel like it

Reply:
I felt